docs/RFCS/20211018_index_recommendation.md

```diff
@@ -0,0 +1,207 @@
+Also, creating an index has a storage cost. Deciding a *fair* cost to associate with creating an index and maintaining an index for database writes is a pending task.
+This is largely dependent on user preference, as some users might prioritize read performance over write performance, and vice versa. To handle this, we can have user settings that allow the user to indicate their preference, which will then affect the cost for writes we use internally.
+These user settings should be specific to each "application name", to deal with the fact that some applications may be more latency-sensitive than others.
+
+Furthermore, creating and removing indexes has significant overhead, so we will use hypothetical indexes instead. There is an existing [prototype PR](https://github.com/cockroachdb/cockroach/pull/66111/) for this. However, these indexes persist to disk, and for our purposes, we only need the indexes in memory.
+We will need to additionally create fake tables that we can tamper with, without interfering with costing on concurrent queries. The implementation idea is to create a struct that wraps `optTable` with additional information pertaining to the table's hypothetical indexes.
+Our hypothetical indexes will be added and removed from this table, that is otherwise identical to the regular table.
+
+Moreover, when recommending the removal of indexes we must be cautious with `UNIQUE` indexes. If a unique index already exists, we have to ensure that we do not remove the unique constraint. This can easily be done by keeping unique indexes.
+For additional flexibility, however, we could consider adding a `UNIQUE WITHOUT INDEX` constraint, which would allow the unique index's removal.
+
+Running the costing algorithm so many times is another hurdle in terms of computational cost.
+We run the algorithm for each statement in *W*, for each statement's candidate set, which is O(cardinality of *W*, squared) time.
+Since the queries we are concerned with are a subset of the statements in *W*, the time complexity is not a tight upper bound. However, it shows that this algorithm has roughly quadratic time complexity, which is quite slow. 
+A way of mitigating this is by only allowing the recommendation algorithm to be run if database utilisation is below a certain threshold, similar to what Azure does [here](https://cockroachlabs.atlassian.net/wiki/spaces/SQLOBS/pages/2252767632/Index+recommendations#:~:text=it%E2%80%99s%20postponed%20if%2080%25%20resource%20utilizated.).
+Another way of mitigating this is by ensuring the sample workload is a meaningful *sample* which is not too large. We do this by limiting the size of the sample workload when we fetch it. 
+If performance continues to be an issue, which is highly applicable for auto-tuning, this functionality can be disabled. Alternatively, a setting can be configured to tune the database less frequently.
+
+## Drawbacks
```

Comment:
nit: some of the things in this section overlap with the previous section. Maybe the sections can be merged?